### PR TITLE
MONGOCRYPT-263 Add RSA signing with macOS

### PR DESCRIPTION
**Pull request by kevinAlbs**

### CODE REVIEW

File: kms-message/src/kms_crypto_apple.c [hunk 1]
```diff
@@ -58,4 +61,83 @@ kms_sha256_hmac (void *unused_ctx,
...
+   if (1 != CFArrayGetCount (out_ref)) {
+      goto cleanup;
+   }
+
+   key_ref = (SecKeyRef) CFArrayGetValueAtIndex (out_ref, 0);
+   data_to_sign_ref = CFDataCreate (NULL, (const uint8_t *) input, input_len);
+   error_ref = NULL;
+   signature_ref =
+      SecKeyCreateSignature (key_ref,
+                             kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256,
```
Comment (kevinAlbs) [inline kms-message/src/kms_crypto_apple.c:126]: Aside, `kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256` is a distinct algorithm (and produces different results). I'm not sure what the difference is (my best guess is that it is the order of signing / hashing).

File: kms-message/src/kms_crypto_apple.c [hunk 2]
```diff
@@ -58,4 +61,83 @@ kms_sha256_hmac (void *unused_ctx,
...
+                           size_t input_len,
+                           unsigned char *signature_out)
+{
+   CFDataRef key_data_ref = NULL;
+   CFDataRef pass_ref = NULL;
+   SecItemImportExportKeyParameters import_params;
+   OSStatus status;
+   /* TODO: I think the expected format should be kSecFormatWrappedPKCS8, but
+    * GCP keys appear to only load for kSecFormatBSAFE. */
+   SecExternalFormat format = kSecFormatUnknown;
```
Comment (kevinAlbs) [inline kms-message/src/kms_crypto_apple.c:86]: I'm still not sure what the deal is here. I left this as `kSecFormatUnknown` for now.

Comment (kevinAlbs) [reply kms-message/src/kms_crypto_apple.c:86]: Per slack discussion, `kSecFormatWrappedPKCS8` is distinct from `PKCS8`.

File: kms-message/src/kms_crypto_apple.c [hunk 3]
```diff
@@ -58,15 +61,83 @@ kms_sha256_hmac (void *unused_ctx,
...
+   if (1 != CFArrayGetCount (out_ref)) {
+      goto cleanup;
+   }
+
+   key_ref = (SecKeyRef) CFArrayGetValueAtIndex (out_ref, 0);
+   data_to_sign_ref = CFDataCreate (NULL, (const uint8_t *) input, input_len);
+   error_ref = NULL;
+   signature_ref =
+      SecKeyCreateSignature (key_ref,
+                             kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256,
```
Comment (markbenvenuto) [inline kms-message/src/kms_crypto_apple.c:126]: The SignatureDigest version expects a digest. The one you chose will do the digest of the message for you.

From the SecKey.h header:
 @constant kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256
    RSA signature with PKCS#1 padding, input data must be SHA-256 generated digest.

  @constant kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256
    RSA signature with PKCS#1 padding, SHA-256 digest is generated from input data of any size.

File: kms-message/src/kms_crypto_apple.c [hunk 4]
```diff
@@ -58,15 +61,83 @@ kms_sha256_hmac (void *unused_ctx,
...
+   }
+   if (1 != CFArrayGetCount (out_ref)) {
+      goto cleanup;
+   }
+
+   key_ref = (SecKeyRef) CFArrayGetValueAtIndex (out_ref, 0);
+   data_to_sign_ref = CFDataCreate (NULL, (const uint8_t *) input, input_len);
+   error_ref = NULL;
+   signature_ref =
+      SecKeyCreateSignature (key_ref,
```
Comment (markbenvenuto) [inline kms-message/src/kms_crypto_apple.c:125]: You need to check the return. You may want to check all the CFDataCreate also.

**Review by markbenvenuto — CHANGES_REQUESTED** [review]

**Review by markbenvenuto — APPROVED** [review]

**Pull request merged**